core/src/main/scala/cats/data/HashSet.scala

```diff
@@ -0,0 +1,840 @@
+    *
+    * @return an iterator that iterates through this set in the reverse order of [[HashSet#iterator]].
+    */
+  final def reverseIterator: Iterator[A] =
+    new HashSet.ReverseIterator(rootNode)
+
+  /**
+    * The size of this set.
+    *
+    * @return the number of elements in this set.
+    */
+  final def size: Int = rootNode.size
+
+  /**
+    * Tests whether the set is empty.
+    *
+    * @return `true` if the set contains no elements, `false` otherwise.
+    */
+  final def isEmpty: Boolean = size == 0
+
+  /**
+    * Tests whether the set is not empty.
+    *
+    * @return `true` if the set contains at least one element, `false` otherwise.
+    */
+  final def nonEmpty: Boolean = !isEmpty
+
+  /**
+    * Apply `f` to each element for its side effects.
+    *
+    * @param f the function to apply to each element.
+    */
+  final def foreach[U](f: A => U): Unit =
+    rootNode.foreach(f)
+
+  /**
+    * Test whether the set contains `value`.
+    *
+    * @param value the element to check for set membership.
+    * @return `true` if the set contains `value`, `false` otherwise.
+    */
+  final def contains(value: A): Boolean =
+    rootNode.contains(value, hash.hash(value), 0)
+
+  /**
+    * Creates a new set with an additional element, unless the element is already present.
+    *
+    * @param value the element to be added.
+    * @return a new set that contains all elements of this set and that also contains `value`.
+    */
+  final def add(value: A): HashSet[A] = {
+    val valueHash = hash.hash(value)
+    val newRootNode = rootNode.add(value, valueHash, 0)
+
+    if (newRootNode eq rootNode)
+      this
+    else
+      new HashSet(newRootNode)
+  }
+
+  /**
+    * Creates a new set with the given element removed from this set.
+    *
+    * @param value the element to be removed.
+    * @return a new set that contains all elements of this set but that does not contain `value`.
+    */
+  final def remove(value: A): HashSet[A] = {
+    val valueHash = hash.hash(value)
+    val newRootNode = rootNode.remove(value, valueHash, 0)
+
+    if (newRootNode eq rootNode)
+      this
+    else
+      new HashSet(newRootNode)
+  }
+
+  /**
+   * Typesafe equality operator.
+   *
+   * This method is similar to [[scala.Any#==]] except that it only allows two [[cats.data.HashSet]]
+   * values of the same element type to be compared to each other, and uses equality provided
+   * by [[cats.kernel.Eq]] instances, rather than using the universal equality provided by
+   * [[java.lang.Object#equals]].
+   *
+   * @param that the [[cats.data.HashSet]] to check for equality with this set.
+   * @return `true` if this set and `that` are equal, `false` otherwise.
+   */
+  def ===(that: HashSet[A]): Boolean = {
+    (this eq that) || {
+      val iterThis = iterator
+      val iterThat = that.iterator
+      this == that
+
+      while (iterThis.hasNext && iterThat.hasNext) {
```

Comment:
is this safe? This implies the iteration order only depends on the `Hash[A]` value and not the order in which they are added. Can we add a comment explaining? I guess specifically about what happens on hash collisions to get a deterministic order here.

Reply:
Yes that is a good point - I have not thought about what happens here when we have a hash collision. We're just using a `Vector` for those.

FWIW, this is a bit of a stopgap while I think about how to implement equality for `BitMapNode` and `CollisionNode`, since that is how `equals` for `HashSet` implemented in the standard library.

Also I'm not sure what that stray `this == that` is doing there 😆

Reply:
@johnynek I have implemented `===` and `equals` for the `Node`s now which should hopefully avoid this issue